src/machine/i2c.go

```diff
@@ -26,6 +26,32 @@ var (
 	errI2CBusError           = errors.New("I2C bus error")
 )
 
+// I2CTargetEvent reflects events on the I2C bus
+type I2CTargetEvent uint8
+
+const (
+	// I2CReceive indicates target has received a byte from the controller.
+	I2CReceive I2CTargetEvent = iota
+
+	// I2CRequest indicates the controller is expecting a byte from the target.
+	I2CRequest
+
+	// I2CFinish indicates the controller has ended the transaction.
+	I2CFinish
+)
+
+// I2CHandler is the type for handling I2C events as an I2C target.
+//
+// Handlers are called in an Interrupt context, so should do minimum
+// processing and not perform any allocations.
+//
+// r contains a byte sent by the controller (when evt is I2CReceive),
+// else 0.
+//
+// The return value should be one or more bytes to return to the
+// controller (when evt is I2CRequest), else it is ignored.
+type I2CHandler func(evt I2CTargetEvent, r byte) []byte
```

Comment:
Appreciate all the work that went into this, awesome stuff!

Just have one question regarding the I2CHandler- it receives one byte, which I presume is the last read. So is this called one time for each byte received? Maybe worth documenting that each byte in the returned byte slice is added to the buffer (which is max 16 length)?

Reply:
Yep - it's one byte at a time for read.  We could aggregate into an array, but then it needs a pre-allocated buffer, so some decision on the max number of permitted bytes from the controller.

To make it device neutral, I could define a constant (`MaxI2CHandlerResponseBytes`?) that is set to the size of the underlying hardware buffer.

Reply:
Yeah I think I'd avoid not writing the bytes over wire immediately (based on intuition alone though). The constant sounds like a good idea to me :+1: